code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm

```diff
@@ -39,24 +40,101 @@
 	core.parent = src
 	RegisterSignal(src, COMSIG_LIVING_WEEDS_ADJACENT_REMOVED, .proc/check_weeds_and_move)
 	RegisterSignal(src, COMSIG_XENOMORPH_CORE_RETURN, .proc/return_to_core)
+	RegisterSignal(src, COMSIG_XENOMORPH_HIVEMIND_CHANGE_FORM, .proc/change_form)
+
+/mob/living/carbon/xenomorph/hivemind/upgrade_possible()
+	return FALSE
+
+/mob/living/carbon/xenomorph/hivemind/updatehealth()
+	health = maxHealth - getFireLoss() - getBruteLoss() //Xenos can only take brute and fire damage.
+	if(health <= 0)
+		setBruteLoss(0)
+		setFireLoss(maxHealth - 1)
+		change_form()
+	med_hud_set_health()
+	if(TIMER_COOLDOWN_CHECK(src, COOLDOWN_HIVEMIND_MANIFESTATION))
+		return
+	update_wounds()
+
+/mob/living/carbon/xenomorph/hivemind/handle_living_health_updates()
+	if(TIMER_COOLDOWN_CHECK(src, COOLDOWN_HIVEMIND_MANIFESTATION))
+		return
+	var/turf/T = loc
+	if(!T || !istype(T))
+		return
+	if(status_flags & INCORPOREAL || locate(/obj/effect/alien/weeds) in T)
+		if((health >= maxHealth) || on_fire) //can't regenerate.
+			updatehealth() //Update health-related stats, like health itself (using brute and fireloss), health HUD and status.
+			return
+		heal_wounds(XENO_RESTING_HEAL)
+		updatehealth()
+		return
+	adjustBruteLoss(20 * XENO_RESTING_HEAL, TRUE)
 
 /mob/living/carbon/xenomorph/hivemind/Destroy()
 	if(!QDELETED(core))
 		QDEL_NULL(core)
 	else
 		core = null
+	upgrade = XENO_UPGRADE_BASETYPE
 	return ..()
 
+/mob/living/carbon/xenomorph/hivemind/change_form()
+	if(status_flags & INCORPOREAL && health != maxHealth)
+		to_chat(src, span_xenowarning("You do not have the strength to manifest yet!"))
+		return
```

Comment:
i don't know how this being tied to the health is that good, i'd prefer if it had proper cooldowns to each situation(when it hits 0 health, switches forms back and etc, this feels way too hard to grasp/control.

Plus from what i see when you get form changed you dont enter any cooldown other than cooldown to transform ? So it seems very easy to just jump in and out.

Reply:
There's nothing hard to understand or control about health.

At zero heath, you get booted to core. You have a visual representation (your health bar) telling you when you will be able to materialise again.

The regen rate is fixed. 

"Plus from what i see when you get form changed you dont enter any cooldown other than cooldown to transform ? So it seems very easy to just jump in and out."

Which is totally the point?